### ISSUE TITLE
About the fid of imagenet64_cond

### ISSUE THREAD

**Issue by encoreus** [apple/ml-tarflow#7]
excuse me, I trained a tarflow model of imagenet64_cond, but I can only get a FID of 6.27, my evaluate_fid.py setting is:

torchrun --standalone --nproc_per_node=4 evaluate_fid.py \
  --data=/tar_flow/data \
  --logdir=runs_original_flow/imagenet64_cond/ \
  --ckpt_file=cond_imagenet_64/imagenet_model_2_768_8_8_0.05.pth \
  --dataset=imagenet \
  --img_size=64 \
  --channel_size=3 \
  --patch_size=2 \
  --channels=768 \
  --blocks=8 \
  --layers_per_block=8 \
  --noise_std=0.05 \
  --nvp \
  --cfg=2.3 \
  --attn_temp=1.0 \
  --batch_size=512 \
  --num_samples=50000 \
  --self_denoising_lr=1.0

so, is there any mistakes in my test setting? or my training is not sufficient? Can you upload the best model of imagenet64 and 128? (since there is only the checkpoint of afhq)

**Comment by Shuangfei** [comment]
Hi, can you share your training script, also what’s the final loss value you get? And sorry that we cannot share the imagenet checkpoints due to licensing issue.

**Comment by encoreus** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，您的邮件已收到，我会尽快处理，谢谢~                                                                       刘犇

**Comment by Doraemonzzz** [comment]
We provide the training curve of cond_imagenet_64, with results as follows:

<img width="1340" alt="Image" src="https://github.com/user-attachments/assets/565042e1-70eb-4914-8054-df3ef8c6cba7" />

<img width="1334" alt="Image" src="https://github.com/user-attachments/assets/28607cff-b040-4181-85bb-9bb997366b34" />

<img width="1340" alt="Image" src="https://github.com/user-attachments/assets/5dad2460-c405-4de6-9eed-db90dae0f49c" />

**Comment by Doraemonzzz** [comment]
We also use the following training script:
```
 torchrun   $code_dir/train.py \
        --config=${config} \
        --logdir=${logdir} \
        --data=${data} \
        --wandb \
        --wandb_project=${wandb_project} \
        --wandb_exp_name=${wandb_exp_name}
```
Specifically, I put the config in yaml, and the config for cond_imagenet_64 is as follows: 
```dataset: imagenet
img_size: 64
channel_size: 3
patch_size: 2
channels: 768
blocks: 8
layers_per_block: 8
noise_std: 0.05
batch_size: 256
epochs: 200
lr: 1e-4
nvp: true
cfg: 0
drop_label: 0.1
sample_freq: 5
logdir: runs/imagenet64-cond
```

**Comment by Shuangfei** [comment]
Thanks. The training curves look pretty normal, so is the online FID curve, which means your checkpoint should be reasonable. I'm not exactly sure what causes the difference but 6.27 is way too high. Maybe try a smaller cfg weight, eg 2.1 instead of 2.3?

**Comment by Doraemonzzz** [comment]
Thank you, we will try other cfg for testing.

**Comment by encoreus** [comment]
I firstly searched for the cfg and got the outcome:

![Image](https://github.com/user-attachments/assets/9f3ab5f0-6440-40fa-967b-8d36eed4d15a)

Then I kept the best cfg and searched for the denosing learning rate:

![Image](https://github.com/user-attachments/assets/21ef451f-0df4-4c6d-b043-8c2ad3cf4add)

It seems that for imagenet64cond, the model I trained reach a limitation at FID 3.93, is this a reasonable value？ Or what other hyper parameter can I tune?

**Comment by Shuangfei** [comment]
Ok this is strange. The optimal cfg weight you found is too small, and the best FID is way too high (and you should be able to achieve this number with even a much smaller model.) The Imagenet128 result looks more reasonable though. Did you double check on the FID real stats and make sure it's correctly computed? The other place that could go wrong is by accidentally setting the `annealed_guidance` flag to False in here https://github.com/apple/ml-tarflow/blob/main/evaluate_fid.py#L77, which will result in the optimal cfg being smaller and also giving worse numbers.

**Comment by encoreus** [comment]
Thanks, I search the cfg and denoise_lr again and match the outcome of imagenet128cond to your paper, but the imagenet64cond is still strange, and I plan to retrain the model.
By the way, when I tune the hypter parameters of imagenet64uncond, I found that denoise_lr is effective but cfg and attn_temp seems valid, so when you get the final fid 18.42 of img64uncond, did you tune all three parameters (cfg, attn_temp, denoise_lr)?

**Comment by encoreus** [comment]
<img width="247" alt="Image" src="https://github.com/user-attachments/assets/40a8815f-371e-4209-904a-0b516a5f4043" />

<img width="258" alt="Image" src="https://github.com/user-attachments/assets/b7b40725-564d-48ac-ac98-e2df64045596" />

**Comment by Shuangfei** [comment]
Sounds good. If you want to retrain, maybe also consider the patch size 4 model in the paper which is faster to converge. For the unconditional model, CFG and attn_temp needs to be jointly searched, with the best cfg and attm_temp sometimes you are able to get slightly better FID by using a denoising_lr that’s slightly less than 1.

**Comment by encoreus** [comment]
Thanks!

**Comment by encoreus** [comment]
Excuse me, could you tell me the range of fid of AFHQ model provided in your checkpoint? I use your checkpoint and yeild a fid about 13.60.

**Comment by Doraemonzzz** [comment]
We trained our model on ImageNet 64x64 with a patch size of 4, and the loss and FID (Fréchet Inception Distance) metrics are shown in the figure below:

![Image](https://github.com/user-attachments/assets/22630eae-e39e-40cd-b6a7-cce439518aca)

![Image](https://github.com/user-attachments/assets/fb79262e-2fa3-475d-983e-b2c6f312d13a)

Is the loss and FID values are reasonable?

**Comment by Doraemonzzz** [comment]
On the other hand, we tested this model, and the results are as follows. Even our best results still have a gap compared to the results reported in the paper.

![Image](https://github.com/user-attachments/assets/013b1de6-c9af-4655-b594-0a92bfca3d6b)

![Image](https://github.com/user-attachments/assets/82113f38-7abc-4807-bf04-beaeb3108509)

**Comment by Shuangfei** [comment]
Thanks for the updates. One thing I just realized the paper forgot to include is that the patch 4, 8x8 depth, 1024 width setting was trained for 300 epochs and a batch size of 480, instead of 200 epochs and batch size 256, and all other hyper parameters are the same as the patch 2 run. Checking our internal run's result, it has a comparable training loss to the one you posted above, but has slightly lower online FID. The FID results for this one is 3.99 with cfg of 1.7 (and we didn't tune the denoising lr). But in general cfg values between [1.3, 2.4] all yield FID below 4.5. 

To make sure there is no unexpected issue, I'll also run a test with the exact patch 4 setting and train for 200 epochs and batch size 256 to see if I get the same result.

**Comment by Shuangfei** [comment]
> Excuse me, could you tell me the range of fid of AFHQ model provided in your checkpoint? I use your checkpoint and yeild a fid about 13.60.

We actually never evaluated the FID of this checkpoint, because there is not much reported result in the literature in this setting. But I'll also run a test on it.

**Comment by encoreus** [comment]
we retrain the img64cond patch4 model, and it's better than previous one. Finally we got a fid 4.42 with cfg=2.0, lr=1.0, is this reasonable? Thanks

<img width="270" alt="Image" src="https://github.com/user-attachments/assets/a0837d22-c948-4b72-aaa6-78953404f98a" />

**Comment by Shuangfei** [comment]
Ok great, this seems to be in the right ballpark. FWIW,  for this model, on our side we get FID 4.19 with CFG 2, and the reported 3.99 FID in the paper is with CFG 1.7,  denosing_lr of 1. This lets me believe that your training runs and evaluations should be correct, but nonetheless there seems to be a slight discrepancy on the exact optimal hyperparamters and best FID numbers you managed to get. For this I don't have much explanation other than variance of the results dependent on the exact environment, and also FID is a finicky metric sensitive to small differences in samples.

**Comment by Y-Sisyphus** [comment]
Excuse me, we have also retrained the img64cond patch size 4 model and achieved an FID score of 4.64 using CFG 2.3 and a denoising learning rate of 1. We would like to verify if the training hyperparameters, such as batch size, learning rate, and noise_std, are consistent with those used for the patch size 2 model.

**Comment by encoreus** [comment]
We trained on 8 A800 GPUS with each 80g memory, with default training setting (batch size 128, learning rate 1e-4, noise_std 0,05). When compute FID, we set cfg=2.0, denoise learing rate=1.0 and sampling batch_size=1024, finally we get a FID of 4.42. We found that all these hyper-parameters vary among different machine.

**Comment by Y-Sisyphus** [comment]
Thanks!
